code/WorkInProgress/DrMelonsStuff.dm

```diff
@@ -114,117 +147,167 @@
+		src.vis_contents -= src.occupant
+		src.occupant = null
+		src.UpdateOverlays(null, "fluid_overlay")
+		src.UpdateOverlays(null, "bath_edge")
+		src.on_reagent_change()
+		if (src.reagents.total_volume)
+			playsound(src.loc, "sound/misc/splash_1.ogg", 70, 3)
+
+		for (var/obj/O in src)
+			O.set_loc(get_turf(src))
+
+	proc/turn_tap(mob/user as mob)
+		src.add_fingerprint(user)
+		if (on)
+			user.visible_message("[user] turns off the bathtub's tap.", "You turn off the bathtub's tap.")
+			playsound(src.loc, "sound/effects/valve_creak.ogg", 30, 2)
+			on = FALSE
+		else
+			if(src.reagents.is_full())
+				boutput(user, "<span class='alert'>The tub is already full!</alert>")
+			else
+				user.visible_message("[user] turns on the bathtub's tap.", "You turn on the bathtub's tap.")
+				playsound(src.loc, "sound/misc/pourdrink.ogg", 60, 4)
+				src.on_reagent_change()
+				on = TRUE
+
+	proc/drain_bathtub(mob/user as mob)
+		src.add_fingerprint(user)
+		if (get_dist(usr, src) <= 1 && !is_incapacitated(usr))
+			if (src.reagents.total_volume)
+				user.visible_message("<span class='notice'>[user] reaches into the bath and pulls the plug.", "<span class='notice'>You reach into the bath and pull the plug.</span>")
+				if (ishuman(usr))
+					var/mob/living/carbon/human/H = usr
+					if(!H.gloves)
+						reagents.reaction(H, TOUCH)
+				playsound(src.loc, "sound/misc/drain_glug.ogg", 70, 1)
+				src.reagents.clear_reagents()
+				src.on_reagent_change()
+
+				var/count = 0
+				for (var/obj/O in src)
+					count++
+					qdel(O)
+				if (count > 0)
+					user.visible_message("<span class='alert'>...and something flushes down the drain. Damn!", "<span class='alert'>...and flush something down the drain. Damn!</span>")
+			else
+				boutput(usr, "<span class='notice'>The bathtub's already empty.</span>")
 
 	relaymove(mob/user as mob, dir)
-		src.eject_user()
-		boutput(user, "<span class='notice'>You get out of the bath.</span>")
-
+		src.eject_occupant(user)
 
 	process()
-		if (src.myuser)
-			if(src.myuser.loc != src)
-				src.myuser.pixel_y = 0
-				src.myuser = null
-				src.overlays = null
-				on_reagent_change()
-
+		if (src.on)
+			src.reagents.add_reagent(src.default_reagent, 100)
+			src.on_reagent_change()
+			if (src.reagents.is_full())
+				src.visible_message("<span class='notice'>As the [src] finishes filling, the tap shuts off automatically.</span>")
+				playsound(src.loc, "sound/misc/pourdrink2.ogg", 60, 5)
+				src.on = FALSE
+		if (src.occupant)
+			if(src.occupant.loc != src)
+				src.occupant.pixel_y = 0
+				src.occupant = null
+				src.on_reagent_change()
 				return
-			if(src.reagents.total_volume > 5)
-				reagents.reaction(src.myuser, TOUCH)
-				src.reagents.trans_to(src.myuser,5,1)
-			else
-				src.reagents.clear_reagents()
-				on_reagent_change()
-
-		return
+			if(src.reagents.total_volume)
+				src.reagents.reaction(src.occupant, TOUCH)
+				src.reagents.trans_to(src.occupant, 5)
```

Comment:
are we like... absorbing the reagents through our skin? the above line should be enough to get skin-pen chems to add to people, and other chems to do whatever other skin-stuff they do.

Reply:
this is preserving the previous behavior (older-diff, line 179) but i am perfectly happy to remove it